src/utils/collections.zig

```diff
@@ -62,3 +62,266 @@ pub fn RecyclingList(
+            self.sort();
+            other.sort();
+            for (self.map.keys(), other.map.keys()) |s, o| {
+                if (s != o) return false;
+            }
+            return true;
+        }
+
+        pub fn put(self: *Self, item: T) !void {
+            try self.map.put(item, {});
+            if (self.max == null or item > self.max.?) {
+                self.max = item;
+            } else {
+                self.is_sorted = false;
+            }
+        }
+
+        pub fn remove(self: *Self, item: T) bool {
+            return self.map.orderedRemove(item);
+        }
+
+        pub fn contains(self: Self, item: T) bool {
+            return self.map.contains(item);
+        }
+
+        pub fn count(self: Self) usize {
+            return self.map.count();
+        }
+
+        pub fn items(self: *Self) []const T {
+            self.sort();
+            return self.map.keys();
+        }
+
+        /// subslice of items ranging from start (inclusive) to end (exclusive)
+        pub fn range(self: *Self, start: ?T, end: ?T) []const T {
+            if (self.count() == 0) return &.{};
+            if (start) |s| if (end) |e| if (e <= s) return &.{};
+            self.sort();
+            var items_ = self.items();
+            if (start) |start_| {
+                // .any instead of .first because uniqueness is guaranteed
+                const start_index = switch (find(T, items_, start_, .any)) {
+                    .found => |index| index,
+                    .after => |index| index + 1,
+                    .less => 0,
+                    .greater => return &.{},
+                    .empty => unreachable, // count checked above
+                };
+                items_ = items_[start_index..];
+            }
+            if (end) |end_| {
+                // .any instead of .last because uniqueness is guaranteed
+                const end_index = switch (find(T, items_, end_, .any)) {
+                    .found => |index| index,
+                    .after => |index| index + 1,
+                    .less => return &.{},
+                    .greater => items_.len,
+                    .empty => unreachable, // count checked above
+                };
+                items_ = items_[0..end_index];
+            }
+            return items_;
+        }
+
+        fn sort(self: *Self) void {
+            if (self.is_sorted) return;
+            self.map.sort(struct {
+                items: std.MultiArrayList(std.AutoArrayHashMap(T, void).Unmanaged.Data).Slice,
+                pub fn lessThan(ctx: @This(), a_index: usize, b_index: usize) bool {
+                    return ctx.items.get(a_index).key < ctx.items.get(b_index).key;
+                }
+            }{ .items = self.map.unmanaged.entries.slice() });
+            self.is_sorted = true;
+        }
+    };
+}
+
+/// binary search that is very specific about the outcome.
+/// only works with numbers
+fn find(
```

Comment:
I'm not sure what you mean by "just one round". This function does not return until either the item is found or until it concludes that the item is not in the list.

My thought when naming this was that "find" is what is relevant to callers. Callers want to find the item in the list. The fact that it's a binary search is just an implementation detail. But I'm ok renaming this to binarySearch since its the obvious approach to find an item in a sorted list.

Reply:
e17030db295a9ccb6661df2701f84aaac27c9d52